test/modules/types/message_test.js

```diff
@@ -0,0 +1,240 @@
+const { assert } = require('chai');
+
+const Message = require('../../../js/modules/types/message');
+
+
+describe('Message', () => {
+  describe('initializeSchemaVersion', () => {
+    it('should ignore messages with previously inherited schema', () => {
+      const input = {
+        body: 'Imagine there is no heaven…',
+        schemaVersion: 2,
+      };
+      const expected = {
+        body: 'Imagine there is no heaven…',
+        schemaVersion: 2,
+      };
+
+      const actual = Message.initializeSchemaVersion(input);
+      assert.deepEqual(actual, expected);
+    });
+
+    context('for message without attachments', () => {
+      it('should initialize schema version to zero', () => {
+        const input = {
+          body: 'Imagine there is no heaven…',
+          attachments: [],
+        };
+        const expected = {
+          body: 'Imagine there is no heaven…',
+          attachments: [],
+          schemaVersion: 0,
+        };
+
+        const actual = Message.initializeSchemaVersion(input);
+        assert.deepEqual(actual, expected);
+      });
+    });
+
+    context('for message with attachments', () => {
+      it('should inherit existing attachment schema version', () => {
+        const input = {
+          body: 'Imagine there is no heaven…',
+          attachments: [{
+            contentType: 'image/jpeg',
+            fileName: 'lennon.jpg',
+            schemaVersion: 7,
+          }],
+        };
+        const expected = {
+          body: 'Imagine there is no heaven…',
+          attachments: [{
+            contentType: 'image/jpeg',
+            fileName: 'lennon.jpg',
+          }],
+          schemaVersion: 7,
+        };
+
+        const actual = Message.initializeSchemaVersion(input);
+        assert.deepEqual(actual, expected);
+      });
+    });
+  });
+
+  describe('upgradeSchema', () => {
+    it('should upgrade an unversioned message to the latest version', async () => {
+      const input = {
+        attachments: [{
+          contentType: 'application/json',
+          data: null,
+          fileName: 'test\u202Dfig.exe',
+          size: 1111,
+        }],
+        schemaVersion: 0,
+      };
+      const expected = {
+        attachments: [{
+          contentType: 'application/json',
+          data: null,
+          fileName: 'test\uFFFDfig.exe',
+          size: 1111,
+        }],
+        schemaVersion: Message.CURRENT_SCHEMA_VERSION,
+      };
+
+      const actual = await Message.upgradeSchema(input);
+      assert.deepEqual(actual, expected);
+    });
+
+    context('with multiple upgrade steps', () => {
+      it('should return last valid message when any upgrade step fails', async () => {
```

Comment:
Hm. Looking at the code above, I'm not sure how this test passes. As far as I can tell, an error thrown by the `upgrade()` function will be swallowed by the `_withSchemaVersion()` method. Must be missing something...

Reply:
Yep, it wouldn’t if there was no guarantee migrations would happen consecutively. I explained how that’s done here: https://github.com/signalapp/Signal-Desktop/pull/2128#discussion_r175508436 😄